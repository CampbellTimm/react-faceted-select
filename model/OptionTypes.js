'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
var OptionTypes = {
    Date: 'date',
    Text: 'text',
    TextOption: 'textoption'
};

exports.default = OptionTypes;