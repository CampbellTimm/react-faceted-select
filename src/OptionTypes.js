export const OptionTypes = {
    Date: 'date',
    Text: 'text',
    TextOption: 'textoption'
};