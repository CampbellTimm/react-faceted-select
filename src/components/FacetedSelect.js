import React from 'react';
import PropTypes from 'prop-types';
import {components} from 'react-select';
import CreatableSelect from 'react-select/lib/Creatable';

const FILTER_SEPARATOR = ':';

class FacetedSelect extends React.Component {

    static propTypes = {
        options: PropTypes.arrayOf(PropTypes.shape({
            label: PropTypes.string.isRequired,
            type: PropTypes.string.isRequired, // TODO RF - May be redundant
            getSuggestions: PropTypes.func
        })).isRequired
    };

    state = {
        inputValue: '',
        value: []
    };

    buildOptions = () => {
        const {options} = this.props;
        const {inputValue} = this.state;
        if (inputValue.includes(FILTER_SEPARATOR)) {
            const key = inputValue.split(FILTER_SEPARATOR)[0];
            const option = options.find(o => {
                return o.label === key
            });
            // TODO RF - if no option is matched what do?
            const suggestions = option.getSuggestions ? option.getSuggestions() : [];
            return suggestions
                .map(suggestedValue => {
                    return {
                        value: suggestedValue,
                        label: suggestedValue,
                        originalOption: option
                    }
                });

        } else {
            return options.map(o => {
                return {
                    value: o.label,
                    label: o.label,
                    getSuggestions: o.getSuggestions
                }
            })
        }
    };

    handleChange = (selectedValues, meta) => {
        const {inputValue} = this.state;
        // TODO RF - handle meta-action "create-option"
        const inputHasSeparator = inputValue && inputValue.includes(FILTER_SEPARATOR);
        if (meta.action === 'remove-value' || meta.action === 'pop-value') {
            this.setState({
                value: selectedValues
            });
        } else if ((meta.action === 'select-option' || meta.action === 'create-option') && inputHasSeparator) {
            // selected a suggested value
            const newSelectedValue = selectedValues[selectedValues.length - 1];
            if (meta.action === 'create-option') {
                // No originalOption available
                newSelectedValue.label = `${newSelectedValue.label}`;
                newSelectedValue.value = `${newSelectedValue.value}`;
            } else {
                newSelectedValue.label = `${newSelectedValue.originalOption.label}: ${newSelectedValue.label}`;
                newSelectedValue.value = `${newSelectedValue.originalOption.value}: ${newSelectedValue.value}`;
            }
            // TODO RF - Call parent component (prop method)
            this.setState({
                value: selectedValues
            });
        } else if (meta.action === 'select-option' && !inputHasSeparator) {
            // selected a suggested key
            const selectedOption = selectedValues[selectedValues.length - 1];
            this.setState({inputValue: `${selectedOption.label}:`})
        } else {
            throw new Error(`Unexpected state for input ${inputValue} and meta ${JSON.stringify(meta)}`);
        }
    };

    handleInputChange = (inputValue) => {
        this.setState({inputValue});
    };

    /*
     * See https://react-select.com/props#replacing-components
     */
    renderCustomInput = (props) => {
        // if type 'date' render date-picker
        return (
            <div>
                <components.Input {...props} />
            </div>
        )
    };

    filterOption = (option, inputValue) => {
        if (!inputValue) return true;
        let searchTerm = inputValue.toLowerCase();
        if (searchTerm.includes(FILTER_SEPARATOR)) {
            searchTerm = searchTerm.split(FILTER_SEPARATOR)[1];
        }
        return option.label.toLowerCase().includes(searchTerm);
    };

    render() {
        const options = this.buildOptions();

        const {inputValue, value} = this.state;

        return (
            <CreatableSelect
                isMulti
                components={{
                    Input: this.renderCustomInput
                }}
                placeholder="Search..."
                isClearable={false}
                closeMenuOnSelect={false}
                filterOption={this.filterOption}
                onChange={this.handleChange}
                options={options}
                onInputChange={this.handleInputChange}
                inputValue={inputValue}
                value={value}
                formatCreateLabel={(inputValue) => inputValue}
            />
        )
    }
}

export default FacetedSelect;