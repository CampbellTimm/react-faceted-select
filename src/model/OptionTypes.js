const OptionTypes = {
    Date: 'date',
    Text: 'text',
    TextOption: 'textoption'
};

export default OptionTypes;