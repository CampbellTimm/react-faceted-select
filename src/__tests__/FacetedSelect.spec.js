import React from 'react';
import FacetedSelect from '../components/FacetedSelect';
import { shallow } from "enzyme";
import OptionTypes from "../model/OptionTypes";

const getCaptureProcessNameSuggestions = () => {
    return [
        'GI-PI-Claims',
        'Life-Claims-RetailClaims'
    ]
};

const getFlowSuggestions = () => {
    return [
        'Life-Claims-V10',
        'BounceBack-Emails-v1'
    ]
};

const options = [
    {
        label:"Capture Process Name",
        type: OptionTypes.TextOption,
        getSuggestions: getCaptureProcessNameSuggestions
    },
    {
        label:"Flow",
        type: OptionTypes.TextOption,
        getSuggestions: getFlowSuggestions
    },
    {
        label:"Capture Correlation ID",
        type: OptionTypes.Text
    }
];

describe('FacetedSelect.js', () => {
    it('should suggest options for keys', () => {
        const wrapper = shallow(<FacetedSelect
            options={options}
        />);
        expect(wrapper).toMatchSnapshot();
    });

    it('should suggest values for selected key', () => {
        const wrapper = shallow(<FacetedSelect
            options={options}
        />);
        wrapper.instance().handleInputChange('Capture Process Name:');
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
    });

    // When key input that matches no options e.g. 'Unknown:'

    // Should call this.props.onSelectOption when adding item and input is complete


});

describe('FacetedSelect #filterOption', () => {
    it('should return true for falsy value', () => {
        expect(FacetedSelect.filterOption({}, false)).toEqual(true);
        expect(FacetedSelect.filterOption({}, undefined)).toEqual(true);
        expect(FacetedSelect.filterOption({}, null)).toEqual(true);
        expect(FacetedSelect.filterOption({}, '')).toEqual(true);
    });

    it('should match case insensitive', () => {
        const option = { label: 'FirSt nAmE' };
        expect(FacetedSelect.filterOption(option, 'first name'))
            .toEqual(true);
    });

    it('should match using token AFTER separator', () => {
        const option = { label: 'John' };
        expect(FacetedSelect.filterOption(option, 'First name:John'))
            .toEqual(true);
    });
});