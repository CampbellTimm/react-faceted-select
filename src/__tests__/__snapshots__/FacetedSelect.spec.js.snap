// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacetedSelect.js should suggest options for keys 1`] = `
<StateManager
  closeMenuOnSelect={false}
  components={
    Object {
      "Input": [Function],
    }
  }
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  filterOption={[Function]}
  formatCreateLabel={[Function]}
  inputValue=""
  isClearable={false}
  isMulti={true}
  onChange={[Function]}
  onInputChange={[Function]}
  options={
    Array [
      Object {
        "getSuggestions": [Function],
        "label": "Capture Process Name",
        "value": "Capture Process Name",
      },
      Object {
        "getSuggestions": [Function],
        "label": "Flow",
        "value": "Flow",
      },
      Object {
        "getSuggestions": undefined,
        "label": "Capture Correlation ID",
        "value": "Capture Correlation ID",
      },
    ]
  }
  placeholder="Search..."
  value={Array []}
/>
`;

exports[`FacetedSelect.js should suggest values for selected key 1`] = `
<StateManager
  closeMenuOnSelect={false}
  components={
    Object {
      "Input": [Function],
    }
  }
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  filterOption={[Function]}
  formatCreateLabel={[Function]}
  inputValue="Capture Process Name:"
  isClearable={false}
  isMulti={true}
  onChange={[Function]}
  onInputChange={[Function]}
  options={
    Array [
      Object {
        "label": "GI-PI-Claims",
        "originalOption": Object {
          "getSuggestions": [Function],
          "label": "Capture Process Name",
          "type": "textoption",
        },
        "value": "GI-PI-Claims",
      },
      Object {
        "label": "Life-Claims-RetailClaims",
        "originalOption": Object {
          "getSuggestions": [Function],
          "label": "Capture Process Name",
          "type": "textoption",
        },
        "value": "Life-Claims-RetailClaims",
      },
    ]
  }
  placeholder="Search..."
  value={Array []}
/>
`;
